### PR TITLE
snap: migrate to `base: core24`

### PR DESCRIPTION
**Pull request by Saviq**

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -3,7 +3,9 @@ adopt-info: mir-test-tools
 summary: Mir test scripts and applications
 description: Mir test scripts and applications
 confinement: strict
-base: core22
+base: core24
+build-base: devel
+grade: devel
```
Comment (AlanGriffiths) [inline snap/snapcraft.yaml]: I *think* if we use snapcraft 8.2 (currently `beta`) we can avoid all this "devel" hacking

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -27,24 +29,34 @@ environment:
...
 apps:
   smoke-test:
     command-chain:
       - bin/wrapper
-      - bin/graphics-core22-wrapper
+      - bin/gpu-2404-wrapper
     command: usr/bin/mir-smoke-test-runner
+    plugs: &plugs
+      - audio-playback
+      - gpu-2404
```
Comment (AlanGriffiths) [inline snap/snapcraft.yaml:41]: We (need to) have `gpu-2404` as a global plug, I don't think we need it here too.

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Consolidating the plugs & slots does avoid some repetition, but makes it less clear why each is required. That might matter if we come to clear things up. (It's a fairly balanced trade off - I can be persuaded.)

File: snap/snapcraft.yaml (hunk 2 above)
Comment (Saviq) [reply snap/snapcraft.yaml:41]: You'd think… but check e.g. [mir-test-tools-22-edge](https://launchpad.net/~mir-team/+snap/mir-test-tools-22-edge):
```
Warning: implicit plug assignment in 'login-session-control', 'opengl', and 'x11'. Plugs should be assigned to the app to which they apply, and not implicitly assigned via the global 'plugs:' stanza which is intended for configuration only.
(Reference: https://snapcraft.io/docs/snapcraft-interfaces)
Warning: implicit slot assignment in 'wayland'. Slots should be assigned to the app to which they apply, and not implicitly assigned via the global 'slots:' stanza which is intended for configuration only.
(Reference: https://snapcraft.io/docs/snapcraft-interfaces)
```

File: snap/snapcraft.yaml (hunk 1 above)
Comment (Saviq) [reply snap/snapcraft.yaml]: But I can't get that for the CI builds…

https://bugs.launchpad.net/snapcraft/+bug/2061517

**Review by Saviq — COMMENTED** [review]
> Consolidating the plugs & slots does avoid some repetition, but makes it less clear why each is required. That might matter if we come to clear things up. (It's a fairly balanced trade off - I can be persuaded.)

You mean rather than in "global" `plugs:`? Yeah, I learned that's not to be used that way… check e.g. [mir-test-tools-22-edge](https://launchpad.net/~mir-team/+snap/mir-test-tools-22-edge):
```
Warning: implicit plug assignment in 'login-session-control', 'opengl', and 'x11'. Plugs should be assigned to the app to which they apply, and not implicitly assigned via the global 'plugs:' stanza which is intended for configuration only.
(Reference: https://snapcraft.io/docs/snapcraft-interfaces)
Warning: implicit slot assignment in 'wayland'. Slots should be assigned to the app to which they apply, and not implicitly assigned via the global 'slots:' stanza which is intended for configuration only.
(Reference: https://snapcraft.io/docs/snapcraft-interfaces)
```

**Comment by AlanGriffiths** [comment]
> You mean rather than in "global" `plugs:`?

No, I mean shared (by anchors) across all the apps

**Comment by Saviq** [comment]
> No, I mean shared (by anchors) across all the apps

Ah. Well, if we could stick to global plugs like we had them, yeah - but because that's not The Way™, the only other option is enumerating them all… and yes, on the one hand we only had e.g. `pulseaudio` apply to the SDL app, not all of them. We might need someone else to convince us one way or the other.

File: snap/snapcraft.yaml (hunk 2 above)
Comment (Saviq) [reply snap/snapcraft.yaml:41]: Except…
https://snapcraft.io/docs/snapcraft-top-level-metadata#heading--plugs-and-slots-for-an-entire-snap

> Plugs and slots for an [interface](https://snapcraft.io/docs/supported-interfaces) are usually configured per-app or per-daemon within snapcraft.yaml. See [Snapcraft app metadata](https://snapcraft.io/docs/snapcraft-app-and-service-metadata) for more details. However, snapcraft.yaml also enables global plugs and slots configuration for an entire snap:

> `plugs`
> Type: dict (optional)
> These plugs apply to all apps …

Comment (AlanGriffiths) [reply snap/snapcraft.yaml:41]: When code and documentation disagree, trust neither

Comment (Saviq) [reply snap/snapcraft.yaml:41]: The documentation was updated to match code.

**Comment by Saviq** [comment]
@AlanGriffiths this can land now, ARM builds will fail until https://github.com/canonical/snapcraft/pull/4786 makes its way into stable.

File: .github/workflows/snap.yml [hunk 3]
```diff
@@ -31,6 +31,7 @@ jobs:
       with:
         architecture: ${{ matrix.platform }}
         snapcraft-token: ${{ secrets.SNAPCRAFT_TOKEN }}
+        snapcraft-channel: edge
```
Comment (AlanGriffiths) [inline .github/workflows/snap.yml:34]: Really?

Comment (Saviq) [reply .github/workflows/snap.yml:34]: Just for a couple of days ;)

**Review by AlanGriffiths — APPROVED** [review]
> Just for a couple of days ;)

OK then

**Pull request merged**